### PR TITLE
Fix for mainz model

### PR DESCRIPTION
**Pull request by ytaous**
Fixes for Mainz/Zcode model and dynamic shape support: 
- STOP_GRADIENT_EDGES - add Tile for Shape
- additional GetShape check
- avoid AllocKind::kReuse when shape size is 0

### CODE REVIEW

File: onnxruntime/core/graph/graph_utils.cc [hunk 1]
```diff
@@ -289,7 +289,7 @@ bool IsSupportedOptypeVersionAndDomain(const Node& node,
                                        const std::string& op_type,
                                        const std::initializer_list<ONNX_NAMESPACE::OperatorSetVersion>& versions,
                                        const std::string& domain) {
-  return (node.OpType() == op_type && !node.Op()->Deprecated() &&
+  return (node.OpType() == op_type && node.Op() && !node.Op()->Deprecated() &&
```
Comment (SherlockNoMad) [inline onnxruntime/core/graph/graph_utils.cc:292]: Did you hit a case where an op doesn't have a OpSchema? 
Is that still a valid model? I wonder whether it should error out and crash the program?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 2]
```diff
@@ -253,7 +253,7 @@ class PlannerImpl {
     // TODO: This should probably be defined to be the equality operator on TensorShapeProto.
     namespace on = ONNX_NAMESPACE;
     int rank1 = shape1.dim_size();
-    if (shape2.dim_size() != rank1) return false;
+    if (rank1 == 0 || shape2.dim_size() != rank1) return false;
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/allocation_planner.cc:256]: This function itself is only for comparing two shapes. if both shape are empty, should they still be treated as "Same" ?

Comment (SherlockNoMad) [reply onnxruntime/core/framework/allocation_planner.cc:256]: Also, shape.dim_size is returning the rank of the shape. a shape proto with value {2,2,0} will return dim_size of 3.  IIUC, you are looking to check if a shape tensor has any 0 as the dim_value?

File: onnxruntime/core/graph/graph_utils.cc (hunk 1 above)
Comment (ytaous) [reply onnxruntime/core/graph/graph_utils.cc:292]: I am getting error on windows test, it has memory access violation due to this node.Op() is null. It's not an issue on linux test somehow. It's the same model "bert_toy_optimized.onnx" loaded by a few UTs, but failed only in one UT - GradientGraphBuilderTest.TrainingSession_BertToy.

File: onnxruntime/core/framework/allocation_planner.cc (hunk 2 above)
Comment (ytaous) [reply onnxruntime/core/framework/allocation_planner.cc:256]: Here we are trying to address the issue of reused buffer. In this case, the dim_size() is 0 in allocation planner, and it just 
select one from freelist_ that matches the shape (also dime_size() == 0). But at execution frame, the shape is {14, 428}, thus 
exception thrown. Therefore we don't want it to be "AllocKind::kReuse", instead to be "AllocKind::kAllocate".
Any other suggestion/idea how to workaround this issue?  
 	
RuntimeError: Error in execution: Non-zero status code returned while running SoftmaxCrossEntropyLoss node. 
Name:'SoftmaxCrossEntropyLoss_1819' Status Message: /bert_ort/ettao/onnx1/onnxruntime/onnxruntime/core/framework/op_kernel.cc:52 
OrtValue* onnxruntime::OpKernelContext::OutputMLValue(int, const onnxruntime::TensorShape&, size_t) status.IsOK() was false. 
Shape mismatch attempting to re-use buffer. {} != {14,428}. Validate usage of dim_value (values should be > 0) and 
dim_param (all values with the same string should equate to the same size) in shapes in the model.

File: onnxruntime/core/graph/graph_utils.cc (hunk 1 above)
Comment (edgchen1) [inline onnxruntime/core/graph/graph_utils.cc:292]: would this be addressed by https://github.com/microsoft/onnxruntime/pull/4761?

Comment (ytaous) [reply onnxruntime/core/graph/graph_utils.cc:292]: Not sure about it, will it? But it won't hurt to make sure node.Op() is not null before calling Deprecated(), this is where I am getting the NPE from.

Comment (ytaous) [reply onnxruntime/core/graph/graph_utils.cc]: I found this code change is no longer applicable to the PR as there's other fix in place. So reverted, thx.

Comment (ytaous) [reply onnxruntime/core/graph/graph_utils.cc]: Also tried the fix in https://github.com/microsoft/onnxruntime/pull/4761 , I can still repro.

File: orttraining/orttraining/core/graph/gradient_builder_base.cc [hunk 3]
```diff
@@ -91,6 +91,8 @@ Status GetShape(const ArgDef& arg_def, std::vector<Dimension>& shape) {
                     "During GetShape, ", arg_def.name, "'s shape is null.");
   const auto& dims = arg_def.type_proto->tensor_type().shape().dim();
   for (auto dim = dims.begin(); dim < dims.end(); dim++) {
+    ORT_RETURN_IF_NOT((*dim).dim_value() > 0 || (*dim).has_dim_param(),
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder_base.cc:94]: nit. dim->dim_value() and dim->has_dim_param() ?

**Pull request merged**